Allow no padding for rolling windows

kmsquire opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #2007
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

kmsquire commented:
FWIW, the test failures are all the same error, which should be fixed, but should have nothing to do with this PR.

dcherian commented:
Thanks @kmsquire I'll try and review this soon.

pep8speaks commented:
Hello @kmsquire! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`xarray/core/rolling.py`](https://github.com/pydata/xarray/blob/bc4ba6b75e04d2c7beb5b393106260e1902169ac/xarray/core/rolling.py):

> [Line 262:1](https://github.com/pydata/xarray/blob/bc4ba6b75e04d2c7beb5b393106260e1902169ac/xarray/core/rolling.py#L262): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (4)







* In the file [`xarray/tests/test_dataset.py`](https://github.com/pydata/xarray/blob/bc4ba6b75e04d2c7beb5b393106260e1902169ac/xarray/tests/test_dataset.py):

> [Line 6303:1](https://github.com/pydata/xarray/blob/bc4ba6b75e04d2c7beb5b393106260e1902169ac/xarray/tests/test_dataset.py#L6303): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

kmsquire commented:
@dcherian Thanks.  I thought I was done, but I'm finding things that aren't working as expected.  I'm working on adding more tests for expected behavior, so marking as WIP.  Will ping back here when I'm finished (maybe today, maybe early next week).

kmsquire commented:
@dcherian Okay, I think this is in good shape.  I added some more tests, and fixed a few more bugs.  Most of the fixes have been squashed back down to the original commit.

I left the second commit separate for now because it's breaking.  Previously, iterating over a rolling window returned only returned blocks present in the original array, and ignored the chunk of `nan`s before the start or after the end of the array.  This meant that the view of the `DataArray` or `Dataset` that was returned for each iteration was potentially a different size.

Here, instead, the iterator was changed so that each returned view matches the corresponding slice of the output of the `construct()` function.  To me, this seems more intuitive, and makes it easier to develop algorithms without having to call construct.  

Other than the fact that it's breaking, the main drawback (and difference with `construct()`) is that the windows themselves are labeled with coordinates along the rolling axis, except for any `nan` values before the start or after the end of the actual data, which do not have labels.  For comparison, in construct, the window dimension does not have any coordinates associated with it by default.

After writing this, I'm wondering if it might be useful to simply drop the coordinates along the rolling axis, so that the behavior matches the behavior of `construct()` even more closely.

I'm open to thoughts/comments/criticisms/suggestions/questions/whatever.

kmsquire commented:
Also, FWIW, the test failure was caused by a problem in zarr/fsspec (https://github.com/intake/filesystem_spec/issues/707), which is fixed in master on fsspec (https://github.com/intake/filesystem_spec/pull/710).  

So it should be fixed here whenever fsspec makes a release.

dcherian reviewed (commented):
I did an initial review and have a few questions. I haven't reviewed `__iter__` yet.

Thanks for doing this @kmsquire it turned out to be harder than expected!

kmsquire commented:
@dcherian Thank you for reviewing.  I've started working through your comments.

kmsquire commented on xarray/core/rolling.py:
```diff
@@ -200,11 +208,61 @@ def _get_keep_attrs(self, keep_attrs):
 
         return keep_attrs
 
+    def _get_rolling_dim_coords(self, all_dims=False) -> Dict[str, Any]:
```
I like the suggestion to call it `_get_output_coords`.  I'll make that change.

Regarding `all_dims`, in some cases, the return value is used in a select statement.  In that case, we only care about the dimensions where the coordinates are different than `self.obj.coords`, such as when there is no padding (this might be the only case).

In other cases, the return value is used in a call to `DataArray` or `Dataset`.  In these cases, we assume that the data has already been reduced to the proper size along each dimension (if applicable), and we wish to construct something akin to the current `DataArray` or `Dataset`, so we need/want all of the dimensions.

I've added a comment to try to clarify this.

kmsquire replied:
I just realized that I've been mixing up "dims" and "coordinates".  The `all_dims` argument should be `all_coords`, and should return the coordinates used to construct the rolling window if `False`, and all coordinates if `True`.

I'll fix the name.  Hopefully it makes more sense now.

kmsquire commented on xarray/core/rolling.py:
```diff
@@ -200,11 +208,61 @@ def _get_keep_attrs(self, keep_attrs):
+
+            return start_offset, end_offset
+
+        for d in dim:
```
Thanks for calling this out, and sorry it isn't clear.

Originally, it was closer to what you suggested here (although I like your version slightly better).

However, we ran into an issue in our own code where we had a second set of coordinates along the dimension without padding--something like
```
In [9]: da
Out[9]: 
<xarray.DataArray (index: 10)>
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
Coordinates:
  * index    (index) int64 0 1 2 3 4 5 6 7 8 9
    time     (index) int64 1 2 3 4 5 6 7 8 9 10
```

In this case, the second dimension (`time`) is not in `self.dim`, but if `index` needs to be shrunk because of no padding, so does `time`.

So the second part of the loop checks the dimensions of `d` (I'm pretty sure `d` can be multidimensional), to see if any of them are used as rolling dimensions, and reduces `d` along those coordinates as well.

(`test_dataarray.test_rolling_construct()` constructs the `DataArray` above and tests this.)

I think this logic is only needed when returning all coordinates (i.e., when this is being used to construct a new DataArray), and not when the return value is being used by `.sel()`.  Perhaps these should be two separate functions.

Regarding `utils.get_slice_offsets`, I originally wrote that expecting to use it in more than one place, but ended up not needing it (and forgot to remove it).  I believe it's also missing the logic I just described.

Either way, I'll try to clean this up a little to make it clearer, probably with something like you suggest above.

kmsquire replied:
Okay, I didn't look closely at your suggestion.  Something close to that should work fine.

kmsquire replied:
This is now implemented with something quite similar to your suggested code above (thanks!).

kmsquire commented on xarray/core/rolling.py:
```diff
@@ -462,15 +570,17 @@ def _counts(self, keep_attrs):
             .construct(rolling_dim, fill_value=False, keep_attrs=keep_attrs)
             .sum(dim=list(rolling_dim.values()), skipna=False, keep_attrs=keep_attrs)
-        )
+        ).sel(output_dim_coords)
```
That's a good question.  I didn't have this in originally, and it wasn't working.  I'll try to figure out why and either fix it or come back with an explanation here.

kmsquire replied:
Fixed--I wasn't making the rolling window correctly a few lines above.

dcherian commented on xarray/core/rolling.py:
```diff
@@ -512,8 +622,11 @@ def _bottleneck_reduce(self, func, keep_attrs, **kwargs):
             values = values[valid]
 
         attrs = self.obj.attrs if keep_attrs else {}
+        output_dim_coords = self._get_rolling_dim_coords()
```
Yes I think we should avoid the sel in general or at least change it to a slice to prevent an unnecessary copy.

Here's how I tested that out.

``` python
a = xr.DataArray([1, 2, 3], dims="x", coords=dict(x=[0, 1, 2]))
print(a.data.flags["OWNDATA"])
print(a.sel(x=slice(None, None)).data.flags["OWNDATA"])
print(a.sel(a.coords).data.flags["OWNDATA"])
```
```
True
False
True  # this means it's independent of `a`
```

We can also defer this to later, and use the other code path if `pad is False`

kmsquire replied:
I think I addressed the other comments.  I'll try to work on this tomorrow.

kmsquire replied:
To address the error that @mathause pointed out, I had to change this to an `isel`, and am passing a dictionary of slices.  It still is probably possible to do something more efficient and remove the call to `pad` above in some cases, but it's a little tedious to work out, and I'm not sure how much more time I can spend on this right now.

kmsquire commented on xarray/core/variable.py:
```diff
@@ -2091,43 +2104,13 @@ def rolling_window(
-                pads[d] = (win - 1, 0)
+        pads = get_pads(dim, window, center, pad)
 
         padded = var.pad(pads, mode="constant", constant_values=fill_value)
```
Done.

mathause reviewed (commented):
Thanks for working on this! I added some comments to the code.

In general I would still like to make something like
```python
rolled = monthly.rolling(center=True, month=n_months, pad_mode="wrap").mean(skipna=False)
```
work. And for this it could be good to think about how we want to name the new argument (e.g. calling it `pad_mode="none"` or `"no_pad"` or whatever might increase its reusability). _However,_ `pad` has so many arguments that it might become a nightmare to thread them all through. So maybe just sticking with `pad=False` is ok.

@dcherian thoughts?

----

The code also introduces a problem, the following no longer works: 

```python
xr.DataArray([1, 2, 3], coords=dict(x=[1, 1, 1])).rolling(x=3).mean()
```
thus you may have to use `isel` instead of `sel` if possible. Therefore I was also unable to test the following (from https://github.com/pydata/xarray/issues/2007#issuecomment-747376047):

```python
monthly.pad(month=n_months, mode="wrap").rolling(center=True, month=n_months, pad=False).mean(skipna=False)
```

dcherian commented:
>  However, pad has so many arguments that it might become a nightmare to thread them all through. So maybe just sticking with pad=False is ok.

We could also do `pad={"mode": "wrap"}` or `pad=True, pad_kwargs={...}` 

I think i prefer `.pad(time=5, mode="wrap").rolling(time=5, pad=False)` but it does require specifying the window length twice. I was going to bring this up at our next dev meeting (tomorrow!)

cc @pydata/xarray

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+            return [arg] * nroll
+        return arg
+
+    arr_args = [to_array(arg) for arg in args]
```
I haven't been very strict about typing things, but in other code, I've found it better to change the name of the variable when the type changes.  You have a better chance of knowing what the type is, and it makes the code easier to reason about.

But I'm willing to change it.

mathause replied:
Good point! Let's keep it as is.

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+ItemOrSequence = Union[T, Sequence[T]]
+
+
+def expand_args_to_num_dims(
```
Renamed to `expand_args_to_dims`.

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+    # dim is now a list
+    nroll = len(dim)
+
+    def to_array(arg):
```
Changed.

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+            f"{name}={val!r}" for name, val in zip(arg_names, arr_args)
+        )
+        raise ValueError(
+            "Arguments must all be the same length. " f"Received {names_vals}."
```
Updated to only print arguments with the wrong length (and to print out the expected length and the length of each incorrect argument).

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+    if is_scalar(dim):
+        for name, arg in zip(arg_names, args):
+            if not is_scalar(arg):
+                raise ValueError(f"Expected {name}={arg!r} to be a scalar like 'dim'.")
```
Done.

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+
+
+def expand_args_to_num_dims(
+    dim: ItemOrSequence[str],
```
Fixed (but this required more changes than just here to appease mypy).

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+
+
+def get_pads(
+    dim: Sequence[str],
```
Changed.

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+    center: Sequence[bool],
+    pad: Sequence[bool],
+) -> Dict[str, Tuple[int, int]]:
+    """Return a mapping from dim to the amount of padding to use at the each end of that dimension
```
Changed.

kmsquire commented on xarray/tests/test_dataarray.py:
```diff
@@ -6657,10 +6710,12 @@ def test_rolling_pandas_compat(center, window, min_periods):
+@pytest.mark.parametrize("window", (2, 3, 4))
 def test_rolling_construct(center, window):
-    s = pd.Series(np.arange(10))
+    count = 10
```
Done (here and a few other places)

kmsquire commented on xarray/core/utils.py:
```diff
@@ -915,3 +915,98 @@ def iterate_nested(nested_list):
+    window: Sequence[int],
+    center: Sequence[bool],
+    pad: Sequence[bool],
+) -> Dict[str, Tuple[int, int]]:
```
Done.

kmsquire commented:
> The code also introduces a problem, the following no longer works:
> ``` 
> xr.DataArray([1, 2, 3], coords=dict(x=[1, 1, 1])).rolling(x=3).mean()
> ```
> thus you may have to use `isel` instead of `sel` if possible. Therefore I was also unable to test the following (from #2007 (comment)):
> 
> ```
> monthly.pad(month=n_months, mode="wrap").rolling(center=True, month=n_months, pad=False).mean(skipna=False)
> ```

Fixed and added a test for this.  The example at the bottom now works.

kmsquire commented:
Mentioned in one of the comments above, but I think I've reached about the amount of time that I can spend on this right now.  If there are other minor changes, please do let me know.  I can also back out the breaking change if desired (although that will probably take some commit surgery).

dcherian commented:
Thanks @kmsquire we can take it from here.

There was some discussion about syntax at the dev meeting today. There were multiple votes in favour of allowing full control of padding in the rolling object itself. So we could have
1. `.rolling(time=5, x=3, pad={"x": {"mode": "wrap"}, "time": False})`, OR
2. `.rolling(time=5, x=3).pad({"x": {"mode": "wrap"}, "time": False})`

@pydata/xarray thoughts?

kmsquire commented:
> There was some discussion about syntax at the dev meeting today. There were multiple votes in favour of allowing full control of padding in the rolling object itself. So we could have
> 
>  1. `.rolling(time=5, x=3, pad={"x": {"mode": "wrap"}, "time": False})`, OR
>  2. `.rolling(time=5, x=3).pad({"x": {"mode": "wrap"}, "time": False})`

FWIW, I think 1 would be more efficient (but perhaps harder to implement).  With this one, the `DataArrayRolling` object is only created once, with all of the required information.

If 2 were implemented, what should the return value of the `.pad()` call be?  Would it call `construct` on the `rolling` object and return a `DataArray`?  Or would it return an updated `DataArrayRolling` object?  Or would it be a new type entirely (e.g., `DataArrayPaddedRolling`)?

dcherian commented:
Your points about 2 being hard to interpret are a good reason to go with 1! I was thinking it would return a DatasetRolling object but that's not the only interpretation as you point out.

dcherian commented on xarray/tests/test_dataarray.py:
```diff
@@ -6561,47 +6584,56 @@ def test_rolling_properties(da):
 @pytest.mark.parametrize("backend", ["numpy"], indirect=True)
-def test_rolling_wrapped_bottleneck(da, name, center, min_periods):
-    bn = pytest.importorskip("bottleneck", minversion="1.1")
+@pytest.mark.parametrize("center", (True, False, None))
```
failing for `center=True` or `pad=False` (I think).

I think there's an issue with constructing `expected`: it doesn't seem like `get_expected_rolling_indices` is correct for all cases.

dcherian replied:
@cmdupuis3 I fixed the merge conflicts but there's still failing tests. We'd appreciate any help you can give here.

cmdupuis3 commented:
Hey all, has there been any movement on this PR, or maybe elsewhere? It seems like a nice feature to me.
